doc/contract-update.md

```diff
@@ -0,0 +1,56 @@
+# Upgrading and Migrating contracts
+
+We consider two approaches:  *Upgrading* existing contracts to a new version but retaining
+state data as-is, and; *migrating* users to a completely new set of contracts.
+
+We would only migrate in the event of a key loss, which prevents upgrading contracts.
+
+## Upgrade contracts approach
+
+Our contracts are already mostly divided into contracts that contain data and contracts that
+contain logic. This is fortuitous for the pattern we are going to use.
+
+`State.sol` contains most of the state, with some in `Shield.sol`.
+This division could be made complete by moving the `proposeBlock` function out of `State.sol`
+and moving the `Shield.sol` data into `State.sol` (there may be a few other minor adjustments
+TBC). `State.sol` will then only contain storage data and getters/setters for the same.
+It will also need modifying so that the addresses of the logic contracts, which it allows to
+set state, can be updated by a registered contract containing an `upgrade` function (see later).
+
+After that is done, we will never upgrade the `State.sol` contract but will upgrade the
+stateless logic contracts. The logic contracts call `State.sol` and thus we avoid the use of
+`delegateCall`. If, in the future, `State.sol` becomes insufficient for our needs, another
+data contract can be deployed alongside, which adds the new storage variables needed.
+
+We will create another contract, which proxies calls to the logic contracts. This avoids users
+having to repoint their applications to the new contracts. The `Proxy.sol` contract will also
+contain the `upgrade` function.
+
+![contract interaction](./contract-upgrade.png)
+
+### upgrade function
+
+This function will, atomically, carry out the following actions:
+
+1. Repoint `State.sol`'s registered logic contracts to the new logic contracts;
+1. Repoint `Proxy.sol` to the new contracts;
```

Comment:
Just a quick note, I think `Proxy.sol` needs to exist for each contract, as all it does is a bit of assembly that points to another contract. So we will need multiple `Proxy.sol` contracts

Reply:
I was imagining one `Proxy.sol` that knows about all of your contracts. I suppose multiple contracts are more flexible if you wanted ever to add an entirely new contract to your fleet though.